File: queue/queue.js
var Queue = function() {
  this.storage = {};
  this.front = 0;
  this.size = 0;
}

Queue.prototype.enqueue = function(value) {
  this.storage[this.size] = value;
  this.size += 1;
}
Queue.prototype.dequeue = function() {
  if(this.size - this.front) {
    var dequeued = this.storage[this.front];
    delete this.storage[this.front];
    this.front += 1;
    return dequeued;
  }
}
Queue.prototype.length = function() {
  return this.size - this.front;
}

File: queue/queue.rb
class Queue
  def initialize
    @input = []
    @output = []
  end

  def enqueue(element)
    @input << element
  end

  def dequeue
    if @output.empty?
      while @input.any?
        @output << @input.pop
      end
    end

    @output.pop
  end
end
